Library/Homebrew/version.rb

```diff
@@ -36,39 +40,49 @@ def self.create(val)
       end.new(val)
     end
 
+    sig { params(val: T.untyped).returns(T.nilable(Token)) }
     def self.from(val)
+      return NULL_TOKEN if val.nil? || (val.respond_to?(:null?) && val.null?)
+
       case val
       when Token   then val
       when String  then Token.create(val)
       when Integer then Token.create(val.to_s)
-      when nil     then NULL_TOKEN
-      else NULL_TOKEN if val.respond_to?(:null?) && val.null?
       end
     end
 
+    sig { returns(T.untyped) }
     attr_reader :value
 
+    sig { params(value: T.untyped).void }
```

Comment:
Could we restrict this perhaps to `Token`, `String`, `Integer`, `nil` or `NullToken`? Would ease some other code, too.

Reply:
I gave this a go, but I came to the conclusion that adding a type to this would require something like Java's generics where a subclass can specify the type of a parent variable. I don't think Sorbet supports this, though

Reply:
1. Added a type for the `value` parameter for `#initialize`
    - `Token` &rarr; `T.nilable(T.any(String, Integer))`
    - `NullToken` has no parameters for `#initialize`
    - `StringToken` &rarr; `String`
    - `NumericToken` &rarr; `T.any(String, Integer)` (converted to `Integer` by `value.to_i`)
1. Added overrides for `attr_reader :value` with the following return types:
    - `NullToken` &rarr; `NilClass` (The class for plain `nil` is `NilClass`)
    - `StringToken` &rarr; `String`
    - `NumericToken` &rarr; `Integer`
1. Replaced `@value = value` with `@value = T.let(value, T.untyped)`, otherwise we get this error for `NullToken` (and a similar error for `StringToken` and `NumericToken`):

```
version.rb:107: Expected NilClass but found T.nilable(T.any(String, Integer)) for method result type https://srb.help/7005
     107 |    attr_reader :value
                           ^^^^^
  Expected NilClass
    version.rb:107: Method value has return type NilClass
     107 |    attr_reader :value
              ^^^^^^^^^^^^^^^^^^
  Got T.nilable(T.any(String, Integer)) originating from:
    version.rb:59:
    59 |      @value = value
              ^^^^^^
```